vllm/multimodal/processing.py

```diff
@@ -212,62 +526,138 @@ def __call__(
+                text,
+                text_matches,
+                mm_items,
+                hf_inputs,
+            )
 
-            for item_idx, orig_item in enumerate(orig_inputs):
-                for match_text, replace_fn in placeholder_replacements.items():
-                    candidates = candidate_placeholders(tokenizer, match_text)
-                    get_replacement_ids = partial(
-                        replace_fn,
-                        orig_item,
-                        processed_inputs,
-                        item_idx,
-                    )
+            token_ids = _encode(tokenizer, text)
+            matched_repls = [match.prompt_repl for match in text_matches]
 
-                    for match_ids in candidates:
-                        # TODO(youkaichao): Don't update new_token_ids
-                        placeholders = apply_placeholders(
-                            new_token_ids,
-                            match_ids,
-                            get_replacement_ids,
-                        )
+        placeholder_ranges = self._find_placeholders(matched_repls, token_ids)
 
-                        if placeholders is not None:
-                            modality_placeholders.append(placeholders)
+        return token_ids, text, placeholder_ranges
 
-            # yapf: disable
-            mm_placeholders[modality] = modality_placeholders  # type: ignore[index]
-            # yapf: enable
+    def apply(
+        self,
+        prompt_text: str,
+        mm_data: MultiModalDataDict,
+        mm_processor_kwargs: Mapping[str, object],
+    ) -> MultiModalInputsV2:
+        tokenizer = self.ctx.tokenizer
+
+        hf_inputs = self._apply_hf_processor(prompt_text, mm_data,
+                                             mm_processor_kwargs)
+        prompt_ids, = hf_inputs.pop("input_ids").tolist()
+        mm_kwargs = MultiModalKwargs(hf_inputs)
+
+        all_prompt_repls = self._bind_prompt_replacements(mm_data)
+
+        # In case HF processor already inserts placeholder tokens
+        all_placeholders = self._find_placeholders(all_prompt_repls,
+                                                   prompt_ids)
+
+        # Otherwise, we insert them ourselves
+        if all_placeholders:
+            prompt_text = _decode(tokenizer, prompt_ids)
+        else:
+            (
+                prompt_ids,
+                prompt_text,
+                all_placeholders,
+            ) = self._apply_prompt_replacements(
+                mm_data,
+                hf_inputs,
+                prompt_ids,
+                all_prompt_repls,
+            )
+
+        mm_placeholders = {
+            modality: [
+                PlaceholderRange(offset=item["offset"], length=item["length"])
+                for item in items
+            ]
+            for modality, items in full_groupby(all_placeholders,
+                                                key=lambda x: x["modality"])
```

Comment:
Hmmm, I mean just make the `key` fixed in `full_groupby` using `groupby_modality = partial(full_groupby, key=lambda x: x["modality"])` in this file.

So that we can just call `groupby_modality(all_placeholders)` etc without line feed for `key`.

Reply:
I see, let me update this.